### PR TITLE
Code refractoring and unit testing for getDeviceTime

### PR DESCRIPTION
**Pull request by SrinivasanTarget**
@Jonahss ping

### CODE REVIEW

**Comment by SrinivasanTarget** [comment]
https://github.com/appium/node-mobile-json-wire-protocol/pull/38

File: lib/commands/general.js [hunk 1]
```diff
@@ -15,7 +15,7 @@ commands.active = async function () {
   }
 };
 
-commands.getDeviceDateAndTime = async function() {
+commands.getDeviceTime = async function() {
```
Comment (imurchie) [inline lib/commands/general.js]: There is one further change needed, in the log message on line 27.

File: README.md [hunk 2]
```diff
@@ -76,7 +76,7 @@ console.log(await driver.getOrientation()); // -> 'LANDSCAPE'
 | `getCookies`                      | 
 | `getCssProperty`                  | 
 | `getCurrentContext`               |
-| `getDeviceDateAndTime`            | 
+| `getDeviceTime`       		    |
```
Comment (imurchie) [inline README.md]: Spacing is weird here.

**Comment by imurchie** [comment]
Are there no tests for this feature?

File: README.md (hunk 2 above)
Comment (SrinivasanTarget) [inline README.md]: Spacing looks fine in my sublime.Not sure what's going wrong here!. Will check and update again

**Comment by SrinivasanTarget** [comment]
No @imurchie Will definitely add test in sometime. Checking out how to add test.

**Comment by imurchie** [comment]
Rerunning the failed tests (which probably don't have anything to do with this change).

**Comment by SrinivasanTarget** [comment]
@imurchie Added tests please review

**Comment by imurchie** [comment]
Another commit broke our CI, so tests are going to fail for now.

But:

As an e2e test this will only work on a real device. It would be better to have a unit test, so that we can stub the calls and make sure that the method functions correctly.

Finally, after having run this locally, it becomes clear that the function needs to be changed a bit. As it stands, the exception thrown by the simulator case is caught immediately, so we end up with `undefined` as the return value. The try/catch block should only be the real device case. And perhaps an exception should be thrown if it fails.

```js
commands.getDeviceTime = async function () {
  log.info('Attempting to capture iOS device date and time');
  if (this.isRealDevice()) {
    try {
      let idevicedate = await fs.which('idevicedate');
      log.info(`Found idevicedate: '${idevicedate}'`);
      let {stdout} = await exec('idevicedate', ['-u', this.opts.udid]);
      return stdout.trim();
    } catch (err) {
      log.errorAndThrow('Could not capture device date and time using libimobiledevice idevicedate. ' +
                     'Libimobiledevice is probably not installed');
    }
  } else {
    log.warn('getDeviceDateAndTime command works only for real devices');
    throw new errors.NotYetImplementedError();
  }
};
```

**Comment by SrinivasanTarget** [comment]
@imurchie Thanks for the detailed comment.I got the changes now. Can you please help me in writing unit test with any example available?

**Comment by SrinivasanTarget** [comment]
@imurchie Can you please review this?

**Comment by imurchie** [comment]
If this is a unit test, let's move it over into the unit tests (https://github.com/appium/appium-ios-driver/blob/master/test/driver-specs.js).

It then does not need to create a session or anything. Just instantiate the driver object, and call the `getDeviceTime` method.

We need to test two conditions: a real device and a simulator. In order to do this, `IosDriver#isRealDevice` needs to be stubbed in order to return `true` or `false` when we want it. For the simulator case, this is enough, and we would expect to get a `NotYetImplementedError` exception thrown.

For the real device case, when our stub of `IosDriver#isRealDevice` returns `true`, we need a little more. First we need to stub `fs.which` so that we don't need to actually call out to the file system. It doesn't matter what that returns (except that it should be a Promise), because we also need to stub out `exec` so we are not actually running the `idevicedate` program.

The `exec` stub can, for one test, return a date string, which we will expect, and in another test it can throw an exception, which we catch and pass along.

An example of mocking (which is like stubbing, and we have some utilities for it) `fs`. It can be adapted for `fs.which()`: https://github.com/appium/appium-adb/blob/master/test/unit/helper-specs.js#L30

An example of mocking `teen_process#exec`: https://github.com/appium/appium-adb/blob/master/test/unit/android-manifest-specs.js#L38

**Comment by imurchie** [comment]
A stab at the successful real device case (showing the whole driver spec file):

```js
// transpile:mocha

import IosDriver from '../';
import chai from 'chai';
import chaiAsPromised from 'chai-as-promised';
import { withMocks } from 'appium-test-support';
import * as teen_process from 'teen_process';
import { fs } from 'appium-support';


chai.should();
chai.use(chaiAsPromised);

describe('driver', () => {
  it('should instantiate class', () => {
    let driver = new IosDriver();
    driver.should.exist;
  });

  describe('getDeviceTime', withMocks({fs, teen_process}, (mocks) => {
    it('should return a valid date string', async () => {
      let udid = 'some-udid';
      let date = new Date().toString();
      mocks.fs.expects('which')
        .once().returns('/path/to/idevicedate');
      mocks.teen_process.expects('exec')
        .once().withExactArgs('idevicedate', ['-u', udid])
        .returns({stdout: date});
      let driver = new IosDriver();
      driver.opts = {udid};

      (await driver.getDeviceTime()).should.equal(date);
    });
  }));
});
```

**Comment by SrinivasanTarget** [comment]
@imurchie Thanks for detailed explanation. It's a great learning for me!
Have included unit testing now.

File: test/driver-specs.js [hunk 3]
```diff
@@ -14,3 +17,35 @@ describe('driver', () => {
...
+      let driver = new IosDriver();
+      driver.opts = {udid};
+
+      (await driver.getDeviceTime()).should.equal(date);
+    });
+    it('should not return a valid date string', async () => {
+      let udid = 'some-udid';
+      let date = new Date().toString();
+      mocks.fs.expects('which')
+        .once().returns('/path/to/idevicedate');
```
Comment (imurchie) [inline test/driver-specs.js]: This test ought to do the case of no `libimobiledevice` installed. That would mean that `exec` would throw an error saying that it couldn't find `idevicedate`.

You can mock this like:

```js
mocks.teen_process.expects("exec")
  .once()
  .throws("ENOENT");
let driver = new IosDriver();
driver.opts = {udid};
await adb.getDeviceTime()
  .should.eventually.be.rejectedWith("Could not capture device date and time");
```

**Comment by imurchie** [comment]
Can you run, on your machine, run `gulp jshint`, and fix the problems that come up?

Then try out the tests you are running by running `gulp once` (or, while you are editing, you can have `gulp watch` running, which will rerun the unit tests each time you save your document).

**Comment by SrinivasanTarget** [comment]
@imurchie please review

**Comment by imurchie** [comment]
CI is failing because you need to add `appium-test-support` to the dependencies. You can run `npm install --save-dev appium-test-support` and then add `package.json` to your commit, and push.

This will allow the tests to use the `withMocks` function.

**Comment by SrinivasanTarget** [comment]
@imurchie Im not able to squash my commits and unfortunately it includes a merge commit as well.Rebase is not helping me.Please help

**Comment by imurchie** [comment]
What happens in the rebase?

**Comment by SrinivasanTarget** [comment]
Interactive rebase hungs while squashing

**Comment by imurchie** [comment]
Is there any output?

**Comment by SrinivasanTarget** [comment]
Nope two or three commits gets squashed and then it leaves conflicts in branch message

**Comment by SrinivasanTarget** [comment]
Now Tests got passed, does test code look fine now?

**Comment by imurchie** [comment]
Yes, the code looks good now. I'm not sure why rebase is failing, though I think it is probably something circular happening in the merge commit.

**Comment by SrinivasanTarget** [comment]
It happened even before the merge commit:(

**Comment by imurchie** [comment]
Hmm. At this point I would normally save the code, rollback, and put it in anew. :(

**Comment by SrinivasanTarget** [comment]
Ok,will do that now.

**Comment by SrinivasanTarget** [comment]
Closing this will raise a new PR:(

**Pull request closed without merging**